<a href="https://colab.research.google.com/github/fhzhkunming/ST-554-HW7/blob/main/HW7_Hui.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ST 554 HW#7 \
Hui Fang\
3/28/2026

# Goal
The purpose of this homework is to practice fitting MLR and logistic regression models (including penalized or regularized models).

# Data
We will use a dataset from the UCI Machine Learning Repository. This data set is about wine quality (we
played around with this data in class at some point I think). You can learn more about the data [here](https://archive.ics.uci.edu/dataset/186/wine+quality).

The data description describes the following variables:

Input variables (based on physicochemical tests)

1 - fixed acidity\
2 - volatile acidity\
3 - citric acid\
4 - residual sugar\
5 - chlorides\
6 - free sulfur dioxide\
7 - total sulfur dioxide\
8 - density\
9 - pH\
10 - sulphates\
11 - alcohol\
Output variable (based on sensory data):\
12 - quality (score between 0 and 10)\
- Rather than try to predict quality, let’s make our target variable for fitting multiple linear regression
type models alcohol.
- For fitting logistic regression type models we’ll use the type of wine as the response variable.

# To Do:
Create a document that goes through your process of reading the data, combining it, manipulating/creating
any variables, and fitting and choosing a final model for both the multiple linear regression modeling and
the logistic regression modeling described below.

## Read in and Combine Data
- Read in the winequality-red.csv and winequality-white.csv files available on the [UCI machine learning repository site](https://archive.ics.uci.edu/dataset/186/wine+quality).


In [8]:
# import pandas
import pandas as pd
# read in data
wine_red = pd.read_csv('winequality-red.csv', sep=';')
wine_white = pd.read_csv('winequality-white.csv', sep=';')
# check the info of the dataset
wine_red.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [7]:
# check the first five rows of the the red wine data
wine_red.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [3]:
# check the basic info
wine_white.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


In [9]:
wine_white.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.0,0.27,0.36,20.7,0.045,45.0,170.0,1.0010,3.00,0.45,8.8,6
1,6.3,0.30,0.34,1.6,0.049,14.0,132.0,0.9940,3.30,0.49,9.5,6
2,8.1,0.28,0.40,6.9,0.050,30.0,97.0,0.9951,3.26,0.44,10.1,6
3,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6
4,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6


- Combine these two datasets and create a new variable that represents the type of wine (red or white)\
The two data sets have identical structure but have different number of observations, so I combined them and create a new variable 'type' to keep track of which wine is which.

In [11]:
# combine the two datasets and create a new variable 'type' to keep track of which wine is which
wine_red['type'] = 'red'
wine_white['type'] = 'white'
wine = pd.concat([wine_red, wine_white], axis=0, ignore_index=True)

In [12]:
wine.info()  # check dataset info of the combined data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         6497 non-null   float64
 1   volatile acidity      6497 non-null   float64
 2   citric acid           6497 non-null   float64
 3   residual sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free sulfur dioxide   6497 non-null   float64
 6   total sulfur dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  type                  6497 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


In [13]:
wine.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,type
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,red
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5,red
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5,red
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6,red
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,red


Rename some of the variables by replacing spaces with underscores for easier modeling

In [15]:
# rename columns by replacing spaces with underscores
wine = wine.rename(columns = lambda x: x.replace(' ', '_'))
# check the info of the renamed dataset
wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed_acidity         6497 non-null   float64
 1   volatile_acidity      6497 non-null   float64
 2   citric_acid           6497 non-null   float64
 3   residual_sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free_sulfur_dioxide   6497 non-null   float64
 6   total_sulfur_dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  type                  6497 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


## Split the Data
- Split up the data set into a training and test set. For this, use stratified sampling to make sure that you have a similar proportion of white and red wines in the training and test sets. This can be done with the `train_test_split()` function ([see the help](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)).

In [67]:
# import modules needed
import numpy as np
import seaborn as sns
from sklearn import linear_model
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression, LassoCV, Lasso, RidgeCV
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import ElasticNetCV

In [29]:
# split the data using stratified sampling
X_train, X_test, y_train, y_test = train_test_split(
    wine.drop(['alcohol','type'], axis= 1),
    wine['alcohol'],                # use 'alcohol' as response
    test_size = 0.2,                # 80/20 split
    random_state = 42,              # set random state for reproducibility
    stratify = wine['type'])        # stratified sampling based on 'type'

## Regression Task (alcohol as Response)


### Train Models

- Fit four different multiple linear regression models.\
    – At least one should include interaction terms\
    – At least one should include some polynomial terms (you may want to standardize your predictors but that is up to you)\
    – Use CV to select your best MLR model

- Standardize variables. \
I first compute the mean and standard deviation for each numeric predictor using the training data. Next, I standardize these predictors by subtracting their mean and dividing by their standard deviation so that each has mean 0 and standard deviation 1. I then save these training-set means and standard deviations so I can apply the same transformation to the test set.

In [87]:
# calculate the means of numeric variables
means = X_train.apply(np.mean, axis= 0)
means


,0
fixed_acidity,8.203303e-18
volatile_acidity,-1.367217e-17
citric_acid,-2.734434e-17
residual_sugar,3.281321e-17
chlorides,-1.093774e-17
free_sulfur_dioxide,1.367217e-17
total_sulfur_dioxide,1.093774e-17
density,3.828208e-17
pH,2.734434e-18
sulphates,6.836086e-18


In [88]:
# calculate the standard deviations of numeric variables
stds = X_train.apply(np.std, axis = 0)
stds

,0
fixed_acidity,1.0
volatile_acidity,1.0
citric_acid,1.0
residual_sugar,1.0
chlorides,1.0
free_sulfur_dioxide,1.0
total_sulfur_dioxide,1.0
density,1.0
pH,1.0
sulphates,1.0


In [89]:
X_train = X_train.apply(lambda x: (x - np.mean(x))/np.std(x), axis = 0)
X_train

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,quality
4110,-0.780839,-0.235499,1.194170,1.229111,0.054220,2.108751,1.749101,0.623768,-0.174702,0.465371,0.201651
3062,0.833361,-0.903661,1.194170,-0.887164,-0.142398,-1.098649,0.374488,-0.953600,-0.986520,-0.480979,0.201651
5236,1.140827,-0.235499,0.150449,-0.125305,-1.069314,-0.423407,0.198256,0.020657,-0.611835,-0.278190,1.341998
497,-0.012173,0.007469,0.011287,-0.612048,0.953048,0.701997,-0.048470,0.637023,0.637117,1.749702,-0.938696
826,0.218427,-0.417725,0.150449,-0.654374,-0.170487,-1.492540,-1.898910,0.139954,1.136697,0.735756,1.341998
...,...,...,...,...,...,...,...,...,...,...,...
634,0.525894,0.068211,-0.754109,-0.739025,0.475546,0.870807,-0.242326,0.570747,0.324879,0.330178,-0.938696
4273,-0.012173,0.068211,-0.475783,0.043997,-0.676077,-0.423407,0.074893,-0.443275,-1.798339,0.870949,1.341998
5343,-0.165906,-1.146628,-0.267039,1.969807,-0.142398,-0.085786,0.585967,1.074445,-1.610996,-0.954153,1.341998
3063,0.602761,0.614888,1.194170,0.784693,-0.704165,0.870807,0.621213,-0.688496,-0.362044,-1.765310,2.482344


In [90]:
X_train.apply(np.mean, axis = 0) # means of variable in the training set are close to 0

,0
fixed_acidity,8.203303e-18
volatile_acidity,-1.367217e-18
citric_acid,-2.871156e-17
residual_sugar,1.367217e-17
chlorides,-8.203303e-18
free_sulfur_dioxide,1.367217e-17
total_sulfur_dioxide,1.914104e-17
density,2.050826e-17
pH,3.418043e-18
sulphates,6.836086e-18


In [91]:
X_train.apply(np.std, axis = 0)  # standard deviations are 1

,0
fixed_acidity,1.0
volatile_acidity,1.0
citric_acid,1.0
residual_sugar,1.0
chlorides,1.0
free_sulfur_dioxide,1.0
total_sulfur_dioxide,1.0
density,1.0
pH,1.0
sulphates,1.0


- Fitting multiple linear regression (MLR) models

In [92]:
# full model
cv_full_model = cross_validate(
    LinearRegression(),                     # create a reg oject
    X_train,
    y_train,
    cv = 5,
    scoring = "neg_mean_squared_error")

# reduce model
cv_mlr1 = cross_validate(
    LinearRegression(),
    X_train[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'sulphates']],
    y_train,
    cv=5,
    scoring="neg_mean_squared_error")

# model with interaction
X_train_inter = X_train.copy()
X_train_inter['sulphates_pH'] = X_train_inter['sulphates'] * X_train_inter['pH']

cv_mlr2 = cross_validate(
    LinearRegression(),
    X_train_inter[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'sulphates', 'sulphates_pH']],
    y_train,
    cv = 5,
    scoring = "neg_mean_squared_error")

# model with polynomial terms
X_train_ploy = X_train.copy()
X_train_ploy['pH_sq'] = X_train_ploy['pH'] ** 2

cv_mlr3 = cross_validate(
    LinearRegression(),
    X_train_ploy[['fixed_acidity', 'residual_sugar', 'chlorides', 'pH', 'pH_sq']],
    y_train,
    cv = 5,
    scoring = "neg_mean_squared_error")


- Choose the best model


In [93]:
# calculate the RMSE of each model
rmse_full = np.sqrt(-cv_full_model['test_score'].mean())
rmse_mlr1 = np.sqrt(-cv_mlr1['test_score'].mean())
rmse_mlr2 = np.sqrt(-cv_mlr2['test_score'].mean())
rmse_mlr3 = np.sqrt(-cv_mlr3['test_score'].mean())
print(rmse_full, rmse_mlr1, rmse_mlr2, rmse_mlr3) # display the valule of RMSEs

0.5477767685928062 1.0469192981080253 1.0469015284211374 1.0490857520265529


It looks like the full modle is the best based on RMSE.

- Fit a LASSO model with a set of predictors of your choosing\
    – Use at least five predictors\
    – Use CV to select the tuning parameter

In [94]:
lasso_wine = LassoCV(cv =5, random_state = 10)\
                    .fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide', 'sulphates', 'pH']],
                    y_train)
print(lasso_wine.alpha_) # extracting the selected tuning parameter alpha_

0.00045809558731599537


This is the tuning parameter α selected by LassoCV. The number is the result of the cross-validation process LASSO uses to balance model fit and coefficient shrinkage.

In [95]:
# display the fitted LASSO coefficients
pd.Series(lasso_wine.coef_,
          index=['fixed_acidity', 'residual_sugar', 'chlorides',
                 'total_sulfur_dioxide', 'sulphates', 'pH'])

,0
fixed_acidity,-0.182194
residual_sugar,-0.352672
chlorides,-0.418356
total_sulfur_dioxide,-0.301626
sulphates,0.074792
pH,-0.064504


The coefficients indicate how each predictor is associated with alcohol content after LASSO shrinkage. In this model, higher values of all predictors except `sulphates` are associated with lower alcohol content when the other variables are held constant. `Chlorides` and `residual_sugar` show the strongest negative associations, suggesting they contribute most to lowering predicted alcohol levels. `Sulphates` has a small positive coefficient, indicating a slight positive association with alcohol content relative to the other predictors.

- Fit a Ridge Regression model with a set of predictors of your choosing\
    – Use at least five predictors\
    – Use CV to select the tuning parameter

In [83]:
# set up tuning-partmeter grid
alphas = np.logspace(-3, 3, 100)

# set up a pipeline with RidgeCV
ridge_cv = make_pipeline(
    StandardScaler(),
    RidgeCV(alphas = alphas, cv = 5))

# fit the model
ridge_cv.fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides', 'total_sulfur_dioxide', 'sulphates']],
            y_train)

# extract the chosen alpha
print(ridge_cv.named_steps['ridgecv'].alpha_)

61.35907273413176


Cross-validation selected a tuning parameter of α = 61.36 for the Ridge model, indicating that the model performs best with relatively strong L2 regularization. This level of shrinkage helps stabilize the coefficients in the presence of correlated predictors.

In [84]:
# check the Ridge coefficients
ridge_cv.named_steps['ridgecv'].coef_

array([-0.15416777, -0.34040461, -0.41236854, -0.28322216,  0.05854266])

The Ridge coefficients show that all predictors except `sulphates` are negatively associated with alcohol content, holding the others constant. `Residual sugar` and `chlorides` have the largest negative coefficients, indicating the strongest associations with lower alcohol levels. `Sulphates` has a small positive coefficient, suggesting a slight positive relationship with alcohol content.

- Fit an Elastic Net model with a set of predictors of your choosing\
    – Use at least five predictors\
    – Use CV to select the tuning parameters

In [85]:
wine_en = make_pipeline(
    StandardScaler(),
    ElasticNetCV(cv = 5, random_state = 10,
                l1_ratio = [0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.96, 0.98, 0.99, 1],
                n_alphas = 50))
wine_en.fit(X_train[['fixed_acidity', 'residual_sugar', 'chlorides',
                     'total_sulfur_dioxide', 'sulphates', 'pH']],
            y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('elasticnetcv',
                 ElasticNetCV(cv=5,
                              l1_ratio=[0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.96,
                                        0.98, 0.99, 1],
                              n_alphas=50, random_state=10))])

In [86]:
en = wine_en.named_steps['elasticnetcv'] # extract the ElasticNetCV model from the pipeline
print(en.alpha_)                         # optimal alpha selected by cross‑validation
print(en.l1_ratio_)                      # optimal L1/L2 mixing parameter selected by CV
print(en.coef_)                          # final Elastic Net coefficients

0.006521185398653273
0.1
[-0.17953504 -0.35082452 -0.41475962 -0.2987029   0.0726287  -0.06183139]


Cross-validation selected α = 0.0065 and an l1_ratio of 0.1, indicating that the Elastic Net model favors a penalty dominated by L2 regularization. The resulting coefficients show that all predictors except `sulphates` are negatively associated with alcohol content. `Residual sugar` and `chlorides` have the largest coefficient magnitudes, suggesting the strongest relationships with alcohol content, while `sulphates` shows a small positive association.

### Test Models
- Using your four selected models, compare their performance on the test set.\
    – Do so using RMSE as your model metric\
    – Do so using MAE as your model metric